Zooming for everyone with keyboard keys

Arteben opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What does it change? What other things could this impact? -->

I have added the ability to change the scale for regular players. Previously, these buttons only worked for ghosts. On the client side, key press events are created and then sent to the server, where the scale is adjusted for the corresponding viewport based on the session or playerUid.

Max zoom for players mostly equal 1.

**Media**
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

[video.webm](https://github.com/space-wizards/space-station-14/assets/2144635/d56359d9-4cf3-4948-8fb4-ab357c78ba32)


- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
<!--
Here you can fill out a changelog that will automatically be added to the game when your PR is merged.

Only put changes that are visible and important to the player on the changelog.

Don't consider the entry type suffix (e.g. add) to be "part" of the sentence:
bad: - add: a new tool for engineers
good: - add: added a new tool for engineers

Putting a name after the :cl: symbol will change the name that shows in the changelog (otherwise it takes your GitHub username)
Like so: :cl: PJB
-->

:cl:
- tweak: Most entities can now zoom in their view via the zoom key bindings. Zooming out is still restricted.

deltanedas commented:
zooming out seems like it will be powergamed very hard, gives you a very clear advantage to anyone playing with default zoom so people will do it

QuietlyWhisper commented:
As long as its only zoom in and back to default for living mobs I wouldnt mind this, could be nice

Arteben commented:
> 



> 



> zooming out seems like it will be powergamed very hard, gives you a very clear advantage to anyone playing with default zoom so people will do it

but it also reduces your view, so it's balanced.
This is simply about player convenience. In some situations, the default field of view is convenient, while in others, such as when working in chemistry or botany, it's helpful to have a small field of view. It's just easier to click on pixels that way. Are we trying to make the game inconvenient for everyone and limit their capabilities?

Arteben commented:
> 

By zoom out you mean setting the scale for  2, 3 etc.
оf course a maximum zoom equal 1. And in other cases 4k monitor give you advantage in any case.

OctoRocket commented:
Can normal players zoom out from the default view?

Arteben commented:
> Can normal players zoom out from the default view?

Zoom out - no
Zoom in - yes

metalgearsloth commented:
Can't you just add contenteye to mobs? I'm not entirely sure what the code changes are for.

Arteben commented:
> Can't you just add contenteye to mobs? I'm not entirely sure what the code changes are for.

I understood this. I will rework it.

Arteben commented:
> Can't you just add contenteye to mobs? I'm not entirely sure what the code changes are for.

1) I moved the key press event generation to the client side so that the playerUid can be included in the event.
2) I removed the animation calculation from the client because the same calculation is already happening on the server. When the same values are calculated in different places, it breaks the animation.
3) Added a new flag to the component to ensure that the animation only works for key presses. So, if you change the zoom, it won't try to animate it towards the ZoomTarget value.
4) Minor refactoring and made a little fix for zoom values

ficcialfaint commented on Resources/Prototypes/Entities/Mobs/Player/human.yml:
```diff
@@ -26,6 +26,7 @@
   - type: Alerts
   - type: Actions
   - type: Eye
+  - type: ContentEye
```
I think you can just add the component for `Resources/Prototypes/Entities/Mobs/Species/base.yml` instead of copy pasting it for all the species

Arteben replied:
fixed

metalgearsloth commented:
> > Can't you just add contenteye to mobs? I'm not entirely sure what the code changes are for.
> 
>     1. I moved the key press event generation to the client side so that the playerUid can be included in the event.
> 
>     2. I removed the animation calculation from the client because the same calculation is already happening on the server. When the same values are calculated in different places, it breaks the animation.
> 
>     3. Added a new flag to the component to ensure that the animation only works for key presses. So, if you change the zoom, it won't try to animate it towards the ZoomTarget value.
> 
>     4. Minor refactoring and made a little fix for zoom values

1. is fine though keybinds already have it afaik.
2. _userZoomChangeRequestTimeOut looks like a big code smell, same with IsProcessed. A lot of the shared code was just moved over to server instead when client can already predict it with keybinds and the data is already available. Running it on both means the client can predict eye zoom but now it's server authoritative and feels a lot less responsive. You also replicated the keybinds with an enum as well.
3. That should just be a new method that sets the eye zoom + target zoom rather than a flag.

ElectroJr reviewed (dismissed):
> 1. I moved the key press event generation to the client side so that the playerUid can be included in the event.

> 1. is fine though keybinds already have it afaik.

1 is the opposite of fine, the server should never just read & accept an entity Uid from the a networked event without validating it. In this case, any malicious player that is sitting in the lobby  (has no attached entity) could just start wildly changing the zoom of every other player connected to the server. The right way to get the player uid of a networked event or keybind is via the `ICommonSession.AttachedEntity` of the client that sent the message, which was already being used before, so I don't understand why it was changed. The keybind handler should just stay in shared.

Otherwise I agree with what sloth said.

metalgearsloth commented:
>1 is the opposite of fine, the server should never just read & accept an entity Uid from the a networked event without validating it. In this case, any malicious player that is sitting in the lobby (has no attached entity) could just start wildly changing the zoom of every other player connected to the server.

If we ever had multi-entity inputs (e.g. camera zoom) though I guess that'd probably be a different system

Arteben commented:
I moved back key bindings. Actually I reworked it again. 
1) Added value checks for ZoomTarget in shared method. 
2) Added an enum instead two classes for key bindings.
3) Added method HasContentEyeComp 
4) Add method SetTargetZoomDirectly and implimented in admin smites and in shuttle consoles.
5) I made the zoom animation has only on client side. For the animation end creates event and this send to server then set values for SharedEyeComponent. 

I have only problem with some error
`Unhandled exception. System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.Collections.Generic.Queue`1.Enumerator.MoveNext()
   at Robust.Client.GameStates.ClientGameStateManager.PredictTicks(GameTick predictionTarget) in /media/artem/Work/dot-net/space-station-14/RobustToolbox/Robust.Client/GameStates/ClientGameStateManager.cs:line 428
   at Robust.Client.GameStates.ClientGameStateManager.ApplyGameState() in /media/artem/Work/dot-net/space-station-14/RobustToolbox/Robust.Client/GameStates/ClientGameStateManager.cs:line 371
   at Robust.Client.GameController.Tick(FrameEventArgs frameEventArgs) in /media/artem/Work/dot-net/space-station-14/RobustToolbox/Robust.Client/GameController/GameController.cs:line 545
   at Robust.Client.GameController.<StartupContinue>b__53_0(Object sender, FrameEventArgs args) in /media/artem/Work/dot-net/space-station-14/RobustToolbox/Robust.Client/GameController/GameController.cs:line 210
   at Robust.Shared.Timing.GameLoop.Run() in /media/artem/Work/dot-net/space-station-14/RobustToolbox/Robust.Shared/Timing/GameLoop.cs:line 218
   at Robust.Client.GameController.ContinueStartupAndLoop(DisplayMode mode) in /media/artem/Work/dot-net/space-station-14/RobustToolbox/Robust.Client/GameController/GameController.Standalone.cs:line 133
   at Robust.Client.GameController.GameThreadMain(DisplayMode mode) in /media/artem/Work/dot-net/space-station-14/RobustToolbox/Robust.Client/GameController/GameController.Standalone.cs:line 118
   at Robust.Client.GameController.<>c__DisplayClass94_0.<Run>b__0() in /media/artem/Work/dot-net/space-station-14/RobustToolbox/Robust.Client/GameController/GameController.Standalone.cs:line 85
`

This appear when user too quickly some zoom key down. I can decide it, but it will be "smell code" with time span.

metalgearsloth commented on Content.Server/Movement/Systems/ContentEyeSystem.cs:
```diff
@@ -5,18 +5,25 @@ namespace Content.Server.Movement.Systems;
+    {
         var query = AllEntityQuery<ContentEyeComponent, SharedEyeComponent>();
 
         while (query.MoveNext(out var uid, out var comp, out var eyeComp))
         {
-            if (eyeComp.Zoom.Equals(comp.TargetZoom))
-                continue;
-
-            UpdateEye(uid, comp, eyeComp, frameTime);
+            if (!eyeComp.Zoom.Equals(comp.TargetZoom))
+            {
+                Sawmill.Debug($"set value {comp.TargetZoom} for zoom animation end");
+                eyeComp.Zoom = comp.TargetZoom;
+                Dirty(eyeComp);
+            }
         }
     }
-}
```
This looks very dirty?

Arteben replied:
> This looks very dirty?

Is this question about new code or old code?

metalgearsloth reviewed (changes requested):
I'm gonna be real I think you should just revert the code changes and only leave the yaml changes as I don't think they're improving the code at the moment.

Arteben commented:
> I'm gonna be real I think you should just revert the code changes and only leave the yaml changes as I don't think they're improving the code at the moment.

You can always close this pull request, I won't be offended, maybe.
About the improvements. I disagree with you. Constantly server-side checks should consume more resources than event-based checks. There are problems with the client crash if the button is pressed very often, but I can solve it, only I do not know how this is decided without "smell code" in this project. Can you give me some example? And all the other edits are just refactorings.

metalgearsloth commented:
1. "EndOfTargetZoomAnimation" refreshes everyone's zooms when the event is received rather than the relevant entity.
2. By only updating server at the end you can't update PVS or whatever as the eye zoom changes (whenever we get dynamic PVS range)
3. You don't get the smooth lerp with some of the server-side stuff anymore
4. Vector should just get a clamp method rather than chaining min / max together
5.

Arteben commented:
Add clamping to engine
Sorry? What is it?

metalgearsloth commented:
> @metalgearsloth "Add clamping to engine" Sorry? What is it?

Clamp is just Min and Max combined so rather than just chaining Min / Max together we just need a clamp method.

Arteben commented:
> I think what should happen is:
> 
>     * Add clamping to engine
> 
>     * Keep some of the validation
> 
>     * Remove most of the other code changes (particularly the predicted event and update removal)

It is reverted almost all. Change check method and added clamp method(I hope what I understand right this). There is some a little animation problem for a first setting zoom, but it seems not critical.

ElectroJr commented on Content.Shared/Movement/Systems/SharedContentEyeSystem.cs:
```diff
@@ -95,15 +99,24 @@ protected void UpdateEye(EntityUid uid, ContentEyeComponent content, SharedEyeCo
 
-    private bool CanZoom(EntityUid uid, ContentEyeComponent? component = null)
+    private ContentEyeComponent? HasContentEyeComp(ICommonSession? session, EntityUid? playerUid = null)
     {
-        return Resolve(uid, ref component, false);
+        var uid = session != null ? session.AttachedEntity : playerUid;
+
+        if (uid is EntityUid entityUid
+            && TryComp<ContentEyeComponent>(entityUid, out var ghostComp))
+        {
+            return ghostComp;
+        }
+        else
+        {
+            return null;
+        }
```
You should remove this function and just use TryComp
